### PR TITLE
Backend: Stats API

### PR DESCRIPTION
**Pull request by Thunderblade73**
## What
Grabs and saves stat information from where ever possible. It has 3 sources: Tablist, SbMenu and the "Your Equipment and Stats" menu.
Also expanded SkyblockStats to include all stats (it didn't have double hook, the crop specifc fortune and cold resitance).

## Changelog Technical Details
+ Added Stats API. - Thunderblade73

### CODE REVIEW

File: src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt [hunk 1]
```diff
@@ -1,62 +1,160 @@
...
+    COLD_RESISTANCE("§b❄", " Cold Resistance: §r§b❄(?<value>.*)", ""),
+    WHEAT_FORTUNE("§7☘", "", " §7(?:§m)☘ Wheat Fortune (?<value>.*)"),
+    CARROT_FORTUNE("§7☘", "", " §7(?:§m)☘ Carrot Fortune (?<value>.*)"),
+    POTATO_FORTUNE("§7☘", "", " §7(?:§m)☘ Potato Fortune (?<value>.*)"),
+    PUMPKIN_FORTUNE("§7☘", "", " §7(?:§m)☘ Pumpkin Fortune (?<value>.*)"),
+    MELON_FORTUNE("§7☘", "", " §7(?:§m)☘ Melon Fortune (?<value>.*)"),
+    MUSHROOM_FORTUNE("§7☘", "", " §7(?:§m)☘ Mushroom Fortune (?<value>.*)"),
+    CACTUS_FORTUNE("§7☘", "", " §7(?:§m)☘ Cactus Fortune (?<value>.*)"),
+    NETHER_WART_FORTUNE("§7☘", "", " §7(?:§m)☘ Nether Wart Fortune (?<value>.*)"),
+    COCOA_BEANS_FORTUNE("§7☘", "", " §7(?:§m)☘ Cocoa Beans Fortune (?<value>.*)"),
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt:75]: You forgot Sugar Cane Fortune 🥺

File: src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt [hunk 2]
```diff
@@ -1,62 +1,160 @@
...
     ;
 
+    var lastKnowValue: Double
```
Comment (lunaynx) [inline src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt]:
```suggestion
    var lastKnownValue: Double
```
(also change it in other places where it's referenced)

**Review by lunaynx — CHANGES_REQUESTED** [review]

**Comment by ItsEmpa** [comment]
not needed here but we could probably also get health, defense and intel from action bar?

**Comment by Thunderblade73** [comment]
> not needed here but we could probably also get health, defense and intel from action bar?

We can also get speed from the player enttiy (same as sba did it),
But this should be done in a later PR. Will add an todo for that.

**Comment by lunaynx** [comment]
> We can also get speed from the player enttiy (same as sba did it)

There's a little caveat here. Your actual Speed stat (post-cap) can technically go up to at least 600, but the actual walk speed reported to the client will never be over 500% of normal walk speed.

The distinction may be relevant for some items such as the Warden Helmet (though I don't know how you'd get 600 speed cap without using both Racing Helmet and Black Cat at the same time).

**Comment by NetheriteMiner** [comment]
> > We can also get speed from the player enttiy (same as sba did it)
> 
> There's a little caveat here. Your actual Speed stat (post-cap) can technically go up to at least 600, but the actual walk speed reported to the client will never be over 500% of normal walk speed.
> 
> The distinction may be relevant for some items such as the Warden Helmet (though I don't know how you'd get 600 speed cap without using both Racing Helmet and Black Cat at the same time).

1. Young dragon armor (but can't be combined with racing helm)
2. By accounting for it with the player entity, you don't need to account for the over 500 thing because it's always right, which is why...
3. SBA's algorithm is already correct, so we don't need to improve on it (in terms of speed cap accuracy, idk about anything else)
4. For getting stuff from the action bar, maybe you can use data.ActionBarStatsData?

**Comment by lunaynx** [comment]
> 1. Rancher's boots

You're several months out of date, they patched that ages ago. Rancher's Boots no longer lets you exceed your normal speed cap of 400 if you aren't using it in combination with items that raise your speed cap actively equipped.

> 2. By accounting for it with the player entity, you don't need to account for the over 500 thing because it's always right, which is why...

It depends on what your goal is.
* Get the player's actual walk speed
  * Yeah, just get it from the entity, that will be accurate in pretty much all cases.
* Get the value of the player's "Speed" SkyBlock stat
  * The entity solution can technically return a wrong value in a few niche cases as explained before.

File: src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt [hunk 3]
```diff
@@ -1,62 +1,161 @@
...
+    MENDING("§a☄", " Mending: §r§a☄(?<value>.*)", " §a☄ Mending §f(?<value>.*)"),
+    TRUE_DEFENCE("§7❂", " True Defense: §r§f❂(?<value>.*)", " §f❂ True Defense §f(?<value>.*)"),
+    SWING_RANGE("§eⓈ", " Swing Range: §r§eⓈ(?<value>.*)", " §eⓈ Swing Range §f(?<value>.*)"),
+    SPEED("§f✦", " Speed: §r§f✦(?<value>.*)", " §f✦ Speed §f(?<value>.*)"), // TODO add the way sba did get it (be careful with 500+ Speed)
+    SEA_CREATURE_CHANCE("§3α", " Sea Creature Chance: §r§3α(?<value>.*)", " §3α Sea Creature Chance §f(?<value>.*)"),
+    MAGIC_FIND("§b✯", " Magic Find: §r§b✯(?<value>.*)", " §b✯ Magic Find §f(?<value>.*)"),
+    PET_LUCK("§d♣", " Pet Luck: §r§d♣(?<value>.*)", " §d♣ Pet Luck §f(?<value>.*)"),
+    FISHING_SPEED("§b☂", " Fishing Speed: §r§b☂(?<value>.*)", " §b☂ Fishing Speed §f(?<value>.*)"),
+    DOUBLE_HOOK_CHANCE("§9⚓", " Double Hook Chance: §r§9⚓(?<value>.*)", ""),
+    BONUS_PEST_CHANCE("§2ൠ", " Bonus Pest Chance: §r§2ൠ(?<value>.*)", " §2ൠ Bonus Pest Chance §f(?<value>.*)"),
```
Comment (ILike2WatchMemes) [inline src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt]:
```suggestion
    BONUS_PEST_CHANCE("§2ൠ", " (?:§r§7§m)?Bonus Pest Chance: (?:§r§2)?ൠ(?<value>.*)", " (?:§7§m|§2)ൠ Bonus Pest Chance (?:§f)?(?<value>.*)"),
```
account for disabled

File: src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt [hunk 4]
```diff
@@ -1,62 +1,161 @@
...
+    ALCHEMY_WISDOM("§3☯", " Alchemy Wisdom: §r§3☯(?<value>.*)", " §3☯ Alchemy Wisdom §f(?<value>.*)"),
+    CARPENTRY_WISDOM("§3☯", " Carpentry Wisdom: §r§3☯(?<value>.*)", " §3☯ Carpentry Wisdom §f(?<value>.*)"),
+    RUNECRAFTING_WISDOM("§3☯", " Runecrafting Wisdom: §r§3☯(?<value>.*)", " §3☯ Runecrafting Wisdom §f(?<value>.*)"),
+    SOCIAL_WISDOM("§3☯", " Social Wisdom: §r§3☯(?<value>.*)", " §3☯ Social Wisdom §f(?<value>.*)"),
+    TAMING_WISDOM("§3☯", " Taming Wisdom: §r§3☯(?<value>.*)", " §3☯ Taming Wisdom §f(?<value>.*)"),
+    MINING_SPEED("§6⸕", " Mining Speed: §r§6⸕(?<value>.*)", " §6⸕ Mining Speed §f(?<value>.*)"),
+    BREAKING_POWER("§2Ⓟ", "", " §2Ⓟ Breaking Power §f(?<value>.*)"),
+    PRISTINE("§5✧", " Pristine: §r§5✧(?<value>.*)", " §5✧ Pristine §f(?<value>.*)"),
+    FORAGING_FORTUNE("§☘", " Foraging Fortune: §r§6☘(?<value>.*)", " §6☘ Foraging Fortune §f(?<value>.*)"),
+    FARMING_FORTUNE("§6☘", " §r§7(?:§m)?Farming Fortune: ☘(?<value>.*)", " §7(?:§m)☘ Farming Fortune (?<value>.*)"),
```
Comment (ILike2WatchMemes) [inline src/main/java/at/hannibal2/skyhanni/data/model/SkyblockStat.kt]:
```suggestion
    FARMING_FORTUNE("§6☘", " (?:§r§7§m)?Farming Fortune: (?:§r§6)?☘(?<value>.*)", " (?:§7§m|§6)☘ Farming Fortune (?:§f)?(?<value>.*)"),
```
account for enabled

**Review by ILike2WatchMemes — CHANGES_REQUESTED** [review]

**Review by CalMWolfs — APPROVED** [review]
code lgtm

**Pull request merged**